### PR TITLE
Better map for signals

### PR DESCRIPTION
**Pull request by to266**
- Separates `map` into two functions that can later be replaced for lazy signals
- As a bonus, running the `map` in parallel (threaded, not multi-process manner) was trivial to add.
- Since I'm making things parallel, now `model.as_signal` can be run in threads as well
## TODO:
- [x] threaded tests to ensure identical behaviour
- [x] threaded tests for `as_signal`

### CODE REVIEW

**Comment by to266** [comment]
Ready for review

File: hyperspy/model.py [hunk 1]
```diff
@@ -410,6 +410,14 @@ def as_signal(self, component_list=None, out_of_range_to_nan=True,
         show_progressbar : None or bool
             If True, display a progress bar. If None the default is set in
             `preferences`.
+        out : {None, BaseSignal}
+            The signal where to put the result into. Convenient for parallel
+            processing. If None (default), creates a new one. If passed, it is
+            assumed to be of correct shape and dtype and not checked.
+        threaded : bool, int
```
Comment (francisco-dlp) [inline hyperspy/model.py]: Why not `parallel` or something of this sort that will be more telling to non-programmers?

File: hyperspy/signal.py [hunk 2]
```diff
@@ -3217,6 +3218,9 @@ def map(self, function,
         show_progressbar : None or bool
             If True, display a progress bar. If None the default is set in
             `preferences`.
+        threaded : bool
```
Comment (francisco-dlp) [inline hyperspy/signal.py]: As above

**Review by francisco-dlp — CHANGES_REQUESTED** [review]
Apart from the minor component this looks very good to me, except that

* [ ] The new feature needs to be documented in the User Guide
* [ ] All calls to `map` in the rest of the code should have the extra `threaded` keyword.

By the way, would it make sense to add `threaded` to `preferences.General`?

**Comment by to266** [comment]
Added as requested

**Comment by francisco-dlp** [comment]
All looks good to me. Could you update your branch with RELEASE_next_patch to see if Appveyor and travis are happy too?

**Comment by to266** [comment]
I just did!

**Comment by francisco-dlp** [comment]
Tests are failing...

**Comment by to266** [comment]
Yep, though I can't reproduce this locally. I'll hopefully figure it out on the next "break"

**Comment by francisco-dlp** [comment]
After upgrading anaconda (`conda update anaconda`) I am able to reproduce the issue locally.

**Comment by to266** [comment]
Seems to have worked. At the same time added notebook progressbar to be used if possible (have been using this for a long time, just never PRed..)

**Review by francisco-dlp — APPROVED** [review]

**Pull request merged**